File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/domain/controllers/commands/get_last_speech.py
# nvdaMcpBridge domain -- GetLastSpeechHandler: the most recent speech sequence.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
#
# ROLE: command handler for `getLastSpeech`.

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .... import protocol
from .command_handler import CommandHandler
from .wallclock import format_wallclock

if TYPE_CHECKING:
	from .session_context import SessionContext


class GetLastSpeechHandler(CommandHandler):
	def execute(self, ctx: SessionContext, request: protocol.Request) -> Any:
		text, index = ctx.speech_buffer.get_last()
		return protocol.LastSpeechResult(
			text=text,
			index=index,
			# The coordinate that places this utterance on the log's timeline
			# (spec 0021); 0 for the sentinel, which no utterance ever occupies.
			logPosition=ctx.speech_buffer.log_position_at(index),
			# Empty for the sentinel, which was never emitted (spec 0028).
			emittedAt=format_wallclock(ctx.speech_buffer.time_at(index)),
		)

File: bridges/voiceover/Sources/VoiceOverBridgeDomain/Controllers/Commands/GetLastSpeech.swift
// ROLE: controller -- `getLastSpeech`: the most recent utterance, with no params.
// BUILT BY: Registry. READS: the session's SpeechBuffer.
// An empty buffer answers with the index-0 sentinel, empty text and empty `emittedAt`, not an error.

import ScreenReaderWire

public final class GetLastSpeechHandler: CommandHandler {
	public init() {}

	public func execute(_ context: SessionContext, _ request: Request) throws -> any Encodable {
		let (utterance, index) = try context.speechBuffer().last()
		return LastSpeechResult(
			text: utterance.text,
			index: index,
			// VoiceOver has no log journal to position into; see Observation.
			logPosition: 0,
			emittedAt: Wallclock.format(utterance.emittedAt)
		)
	}
}
